sockeye/inference.py

```diff
@@ -862,17 +897,27 @@ def _get_inference_input(self,
         for j, trans_input in enumerate(trans_inputs):
             num_tokens = len(trans_input)  # includes eos
             max_output_lengths.append(self._get_max_output_length(num_tokens))
-            source[j, :num_tokens, 0] = tokens2ids(trans_input.tokens, self.source_vocabs[0])
+            num_tokens = num_tokens if not trans_input.source_prefix else num_tokens + len(trans_input.source_prefix.split())
+            if not trans_input.source_prefix:
+                source[j, :num_tokens, 0] = tokens2ids(trans_input.tokens, self.source_vocabs[0])
+            else:
+                #adding source prefix to every chunk
+                source[j, :num_tokens, 0] = tokens2ids(trans_input.source_prefix.split() + \
+                    trans_input.tokens, self.source_vocabs[0])
 
             factors = trans_input.factors if trans_input.factors is not None else []
             num_factors = 1 + len(factors)
             if num_factors != self.num_source_factors:
                 logger.warning("Input %d factors, but model(s) expect %d", num_factors,
                                self.num_source_factors)
-            for i, factor in enumerate(factors[:self.num_source_factors - 1], start=1):
-                # fill in as many factors as there are tokens
-
-                source[j, :num_tokens, i] = tokens2ids(factor, self.source_vocabs[i])[:num_tokens]
+            if not trans_input.source_prefix or not trans_input.source_prefix_factors: #no source prefix during inference
+                for i, factor in enumerate(factors[:self.num_source_factors - 1], start=1):
+                    # fill in as many factors as there are tokens
+                    source[j, :num_tokens, i] = tokens2ids(factor, self.source_vocabs[i])[:num_tokens]
+            else:
```

Comment:
instead of the if-else, could you maybe use`itertools.chain` ?

Reply:
I found using `intertools.chain` a bit complicate here because we need a merge of two separate List: `factor` and `source_prefix_factor` in an one by one mapping (for `intertools.chain` it is about concatenating the twos, which is not what we need here).

Reply:
Yes you are right